### PR TITLE
[WIP] Fixes #26676 - improve setting initialization

### PR DESCRIPTION
**Pull request by ares**
reopening of https://github.com/theforeman/foreman/pull/6708 all info and reasons there, katello tests will be broken so I need to first find out why

### CODE REVIEW

File: app/models/setting.rb [hunk 1]
```diff
@@ -242,9 +243,16 @@ def self.create_existing(s, opts)
...
       to_update = Hash[opts.select { |k, v| attrs.include? k }]
       to_update[:value] = readonly_value(s.name.to_sym) if s.has_readonly_value?
-      s.update(to_update)
+      # default is converted to yaml so we need to convert the yaml here too,
+      # in order to check, if the update of default is needed
+      # if it is the same, we don't try to update default, it would trigger
+      # update query for every setting
+      to_update.delete(:default) if to_update[:default].to_yaml.strip == s[:default]
+      s.attributes = to_update
+      s.save if s.changed?
```
Comment (ezr-ondrej) [inline app/models/setting.rb:252]: I believe `Rails` is doing exactly that (calling `changed?`) internally even on `update`, so I believe the deleting the default is enough here.

File: app/models/setting.rb [hunk 2]
```diff
@@ -271,7 +279,7 @@ def self.cache
 
   def self.load_defaults
     # We may be executing something like rake db:migrate:reset, which destroys this table; only continue if the table exists
-    Setting.first rescue return false
+    Setting.table_exists? rescue return false
```
Comment (ezr-ondrej) [inline app/models/setting.rb:282]: I believe the rescue is never going to be hitted now. I believe only `Setting.table_exists?` should be kept in this method, if you wouldn't want to use my suggestion above.

**Review by ezr-ondrej — CHANGES_REQUESTED** [review]
I love optimizing the `Setting` load as it bugs me for some time.
Big :+1: for me!

I'm just bit worried about the class variable cache, wouldn't it be better to actually change the logic in ancestors to filter unchanged values prior to calling the `create`?
I know it is more work as we would need to change all ancestors, but the class variable seems to be one purposed for application start and is kept around in memory for whole application run, without being reused and if it would be used, it could cause problems as it keeps old values.

Something like this could maybe replace all ancestors `load_default_settings` methods instead?
```ruby
def self.load_default_settings
  return false unless table_exists?
  dbcache = Hash[Setting.where(:category => self.name).map { |s| [s.name, s] }]
  self.transaction do
    default_settings.compact.each do |s|
       val = s.update(:category => self.name).symbolize_keys
       dbcache.key?(val[:name]) ? create_existing(dbcache[val[:name]) : create!(s)
    end 
  end
  true
end
```

It would hit the load query in cases the Setting actually do not exists, but it is only for case the new setting was added and only once. I believe that would be ok, right?

**Comment by ezr-ondrej** [comment]
Any progress on this one?

**Comment by ares** [comment]
Not in forseeable future :-(  I might get to it in late August when I'm travelling.

**Comment by ezr-ondrej** [comment]
So you wouldn't get mad, if someone else would pick it up and try to push it? :stuck_out_tongue:
Purely hypotheticaly, of course: #6902

**Comment by ares** [comment]
Such person would get in fact a beer from me :-) Thanks, closing this!

**Pull request closed without merging**